// @clutch-sh/ec2/dist/index.js
import RebootInstance from "./reboot-instance";
import ResizeAutoscalingGroup from "./resize-asg";
import TerminateInstance from "./terminate-instance";
const register = () => {
    return {
        developer: {
            name: "Lyft",
            contactUrl: "mailto:hello@clutch.sh",
        },
        path: "ec2",
        group: "AWS",
        displayName: "EC2",
        routes: {
            terminateInstance: {
                path: "instance/terminate",
                displayName: "Terminate Instance",
                description: "Terminate an EC2 instance.",
                component: TerminateInstance,
                requiredConfigProps: ["resolverType"],
            },
            rebootInstance: {
                path: "instance/reboot",
                displayName: "Reboot Instance",
                description: "Reboot an EC2 Instance",
                component: RebootInstance,
                requiredConfigProps: ["resolverType"],
            },
            resizeAutoscalingGroup: {
                path: "asg/resize",
                displayName: "Resize Autoscaling Group",
                description: "Resize an autoscaling group.",
                component: ResizeAutoscalingGroup,
                requiredConfigProps: ["resolverType"],
            },
        },
    };
};
export default register;
