// @clutch-sh/ec2/dist/terminate-instance.js
import React from "react";
import { Accordion, AccordionDetails, Button, ButtonGroup, client, Confirmation, MetadataTable, NotePanel, Resolver, useWizardContext, } from "@clutch-sh/core";
import { useDataLayout } from "@clutch-sh/data-layout";
import { Wizard, WizardStep } from "@clutch-sh/wizard";
const InstanceIdentifier = ({ resolverType }) => {
    const { onSubmit } = useWizardContext();
    const resolvedResourceData = useDataLayout("resourceData");
    const onResolve = ({ results }) => {
        // Decide how to process results.
        resolvedResourceData.assign(results[0]);
        onSubmit();
    };
    return React.createElement(Resolver, { type: resolverType, searchLimit: 1, onResolve: onResolve });
};
const InstanceDetails = () => {
    var _a;
    const { onSubmit, onBack } = useWizardContext();
    const resourceData = useDataLayout("resourceData");
    const instance = resourceData.displayValue();
    const data = [
        { name: "Instance ID", value: instance.instanceId },
        { name: "Name", value: (_a = instance === null || instance === void 0 ? void 0 : instance.tags) === null || _a === void 0 ? void 0 : _a.Name },
        { name: "Account", value: instance.account },
        { name: "Region", value: instance.region },
        { name: "State", value: instance.state },
        { name: "Instance Type", value: instance.instanceType },
        { name: "Public IP Address", value: instance.publicIpAddress },
        { name: "Private IP Address", value: instance.privateIpAddress },
        { name: "Availability Zone", value: instance.availabilityZone },
    ];
    const tags = [];
    if (instance.tags) {
        Object.keys(instance.tags).forEach(key => {
            tags.push({ name: key, value: instance.tags[key] });
        });
    }
    return (React.createElement(WizardStep, { error: resourceData.error, isLoading: resourceData.isLoading },
        React.createElement("strong", null, "Instance Details"),
        React.createElement(MetadataTable, { data: data }),
        tags.length > 0 && (React.createElement(Accordion, { title: "Tags" },
            React.createElement(AccordionDetails, null,
                React.createElement(MetadataTable, { data: tags })))),
        React.createElement(ButtonGroup, null,
            React.createElement(Button, { text: "Back", variant: "neutral", onClick: () => onBack() }),
            React.createElement(Button, { text: "Terminate", variant: "destructive", onClick: onSubmit }))));
};
const Confirm = ({ notes }) => {
    const terminationData = useDataLayout("terminationData");
    const instance = useDataLayout("resourceData").displayValue();
    const data = [
        { name: "Instance ID", value: instance.instanceId },
        { name: "Region", value: instance.region },
    ];
    return (React.createElement(WizardStep, { error: terminationData.error, isLoading: terminationData.isLoading },
        React.createElement(Confirmation, { action: "Termination" }),
        React.createElement(MetadataTable, { data: data }),
        React.createElement(NotePanel, { notes: notes })));
};
const TerminateInstance = ({ heading, resolverType, notes = [] }) => {
    const dataLayout = {
        resourceData: {},
        terminationData: {
            deps: ["resourceData"],
            hydrator: (resourceData) => {
                return client.post("/v1/aws/ec2/terminateInstance", {
                    instance_id: resourceData.instanceId,
                    account: resourceData.account,
                    region: resourceData.region,
                });
            },
        },
    };
    return (React.createElement(Wizard, { dataLayout: dataLayout, heading: heading },
        React.createElement(InstanceIdentifier, { name: "Lookup", resolverType: resolverType }),
        React.createElement(InstanceDetails, { name: "Verify" }),
        React.createElement(Confirm, { name: "Result", notes: notes })));
};
export default TerminateInstance;
